security: only perform strict validation on first certificate.

mberhault opened the pull request:
Fixes #38146.

If a client or node certificate includes intermediates, we used to
apply the same strict validation to the intermediates as well. This
will fail on invalid CommonName.

Now also return error when part of the certificate file could not be
parsed.

Release note: (security): only check CN on first certificate in file

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38163)
<!-- Reviewable:end -->

mberhault commented on pkg/security/pem.go:
```diff
@@ -100,6 +100,9 @@ func PEMToCertificates(contents []byte) ([]*pem.Block, error) {
 		block, contents = pem.Decode(contents)
 		if block == nil {
+			if len(contents) > 0 {
+				return nil, errors.Errorf("block #%d could not be parsed", len(certs))
```
This check is stricter than it used to be. If we still want to make use of the certificates seen so far, we could just log a warning.

bdarnell reviewed (approved):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mberhault)

___
*[pkg/security/pem.go, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38163#-LhLTbC3-MC97FPIB7KX-r1-104:-LhLZMJ727awvNgj7cKs:b-dg5trz) ([raw file](https://github.com/cockroachdb/cockroach/blob/72ae2517e20a070b36b22ccee42d8604ebe74ce3/pkg/security/pem.go#L104)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

This check is stricter than it used to be. If we still want to make use of the certificates seen so far, we could just log a warning.
</blockquote></details>

What motivated this change? What kind of problem can lead to this path being taken? 

This sounds like a good change, but it might be safest to make it a separate commit. (So we can backport the other changes but keep this part for 19.2 only). 


<!-- Sent from Reviewable.io -->

Commit:
security: only perform strict validation on first certificate.

Fixes #38146.

If a client or node certificate includes intermediates, we used to
apply the same strict validation to the intermediates as well. This
will fail on invalid CommonName.

Release note (security): only check CN on first certificate in file

mberhault reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38163)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell)

___
*[pkg/security/pem.go, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38163#-LhLTbC3-MC97FPIB7KX-r1-104:-LhLeHSP8bu8znTEwsuE:bdj2vce) ([raw file](https://github.com/cockroachdb/cockroach/blob/72ae2517e20a070b36b22ccee42d8604ebe74ce3/pkg/security/pem.go#L104)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What motivated this change? What kind of problem can lead to this path being taken? 

This sounds like a good change, but it might be safest to make it a separate commit. (So we can backport the other changes but keep this part for 19.2 only). 
</blockquote></details>

I was poking at the test case that adds junk after the certificate and found the handling had changed. It also seems dependent on the version I'm running (pem validation has had some tweaks over the last two years).
I removed it from this PR (including the entire test case) and will put it on a separate one. I would say we want a warning at the very least.


<!-- Sent from Reviewable.io -->

mberhault commented:
TFTR. Are you ok with backporting to both 19.1 and 2.1?

mberhault commented:
bors r+

bdarnell commented:
Yes, this should be backported to both branches.
